Fix missing terms of services link

oneiros opened the pull request:
Fixes #35069

While the problem was pretty clear, I was not 100% sure how to fix it.

I opted for the following: When no terms are currently in effect take the published version that is next up to become effective. In effect the footer link appears if either of those exist.

This makes sense to me, but feel free to let me know if another solution is preferred or if there is something I have missed.

oneiros committed:
Add and use `TermsOfService::current`...

... to return either the terms that are currently in effect or,
in the case that no terms are in effect yet, the published version
that is next up to become effective.

Fixes #35069

mjankowski commented:
Is the desired change here that a TOS record which has been saved and given a future effective date starts getting shown immediately if there is not an earlier/current TOS present?

So hypothetically I create a TOS today with a September 1, 2025 effective date ... if I did not already have a previous one in place and effective, the app would start showing that one right away?

mjankowski commented:
Thanks -- will give code a quick pass, wanted to make sure I understood goals first.

I will say that it's simultaneously sort of odd to show a TOS before it's actually effective ... but also I totally see the motivation there of wanting to show people the upcoming thing. Hopefully the effective date on top is enough to mitigate any confusion.

mjankowski commented on app/models/terms_of_service.rb:
```diff
@@ -16,6 +16,7 @@
 class TermsOfService < ApplicationRecord
   scope :published, -> { where.not(published_at: nil).order(Arel.sql('coalesce(effective_date, published_at) DESC')) }
   scope :live, -> { published.where('effective_date IS NULL OR effective_date < now()').limit(1) }
+  scope :upcoming, -> { published.reorder(effective_date: :asc).where('effective_date IS NOT NULL AND effective_date > now()').limit(1) }
```
A good follow-up to this PR would be a small refactor of these scopes to keep each one to do fewer things, which would probably simplify some of the combining/reordering/usage of them. I realize this PR is not creating the situation though, so prob fine as-is here.

oneiros commented on app/controllers/api/v1/instances/terms_of_services_controller.rb:
```diff
@@ -15,8 +15,9 @@ def set_terms_of_service
+        TermsOfService.current
       end
     end
+    not_found if @terms_of_service.nil?
```
I actually thought about this, but it is *not* the only place.

mjankowski replied:
I noticed roughly 30 seconds later that it's used in places where the raise would be a problem, so this is fine.

mjankowski commented on app/models/terms_of_service.rb:
```diff
@@ -26,6 +27,10 @@ class TermsOfService < ApplicationRecord
   NOTIFICATION_ACTIVITY_CUTOFF = 1.year.freeze
 
+  def self.current
+    TermsOfService.live.first || TermsOfService.upcoming.first # For the case when none of the published terms have become effective yet
```
Do these need the explicit class?

oneiros replied:
:man_facepalming: 

In my defence, it is getting rather hot around here :hot_face: 

Thanks for catching this!

oneiros committed:
Review feedback

No need to reference class name inside class itself.

Use `travel_to` to fabricate records "in the past".
